spec.md

```diff
@@ -179,17 +179,163 @@ physically present in the final circuit. Refer to the description of the
 instance statement for details on how to instantiate a module
 ([@sec:instances]).
 
+## Optional Groups
+
+Optional groups are named collections of statements inside a module.  Optional
+groups contain optional functionality for a module that can be included or
+excluded without affecting the functionality of a circuit.  Optional groups are
+used to keep verification, debugging, or other collateral, not relevant to the
+operation of the circuit, in a separate area.  Each group can then be optionally
+included in the resulting design.
+
+The `declgroup`{.firrtl} keyword declares an optional group with a specific
+identifier.  An optional group may be declared in either a circuit, and external
+module, or in another optional group declaration.  An optional group's
+identifier must be unique within the current namespace.  I.e., the identifier of
+a top-level group declared in a circuit must not conflict with the identifier of
+a module, external module, or implementation defined module.
+
+Each optional group declaration must include a string that sets the lowering
+convention for that group.  The FIRRTL ABI specification defines supported
+lowering convention.  One such strategy is `"bind"`{.firrtl} which lowers to
+modules and instances which are instantiated using the SystemVerilog
+`bind`{.verilog} feature.
+
+The `group`{.firrtl} keyword defines an optional group.  An optional group may
+only be defined inside a module.  An optional group must reference a group
+declared in the current circuit.  Statements in an optional groups may only read
+from ports or declarations of the current module or the current
+group---statements in a group may not drive components declared outside the
+group.  Declarations of identifiers in a group may not collide with identifiers
+declared outside the group.
+
+The circuit below contains one optional group declaration, `Bar`.  Module `Foo`
+contains a group definition that creates a node computed from a port defined in
+the scope of `Foo`.
+
+``` firrtl
+circuit Foo:
+  declgroup "bind", Bar:  ; Declaration of group Bar
+
+  module Foo:
+    input a: UInt<1>
+
+    group Bar:            ; Definition of group Bar inside module Foo
+      node notA = not(a)
+
+  extmodule Baz:
+    input a: UInt<1>
+    output b: UInt<1>
+    declgroup "bind" Bar: ; Declaration of group Bar for extmodule Baz
```

Comment:
Looking at this, I do think not having the declgroups in extmodules might be cleaner:
```
  extmodule Baz using declgroup Bar:
    input a: UInt<1>
    output b: UInt<1>
```
As then you don't have to specify how two declgroups with different implementation strings interact.

Reply:
Or alternately, maybe we just don't need to know this information?